src/core/lib/layer-manager.js

```diff
@@ -351,7 +351,9 @@ export default class LayerManager {
       // TODO - reimplement viewport change detection (single viewport optimization)
       // TODO - don't set viewportChanged during setViewports?
       if (this.context.viewportChanged) {
-        this._updateLayerStates({viewportChanged: true});
+        for (const layer of this.layers) {
+          this._updateLayer(layer, {viewportChanged: true, somethingChanged: true});
```

Comment:
`somethingChanged`, is this supposed to be removed?

Reply:
somethingChanged is set to indicated that any flag changed. I don't think any layer uses it.

Reply:
It's used by the default `shouldUpdateState` of CompositeLayer.

Reply:
Ok, looks to me `somethingChanged` is referring to any change other than props or data and `Layer.diffProps` considers all cases and sets the `changeFlags`. So instead of setting these flags manually here should we always call `diffProps` , in future if any new flags added and diffProps changed this might cause an issue.

Reply:
> It's used by the default `shouldUpdateState` of `CompositeLayer`.

I can't find a `shouldUpdateState` in `CompositeLayer`: https://github.com/uber/deck.gl/blob/master/src/core/lib/composite-layer.js#L24

Reply:
> Ok, looks to me somethingChanged is referring to any change other than props or data and Layer.diffProps considers all cases and sets the changeFlags. So instead of setting these flags manually here should we always call diffProps , in future if any new flags added and diffProps changed this might cause an issue.

Good discussion. In this case the purpose was to avoid calling `diffProps` so I can't really call it again. 

I think the entire changeFlags system could do with a review:
1) it is not clear to me that we need a `somethingChanged` flag: i.e. if nothing has changed, why are we even calling these functions.
2) The flags that we do have should be better documented.
3) A better implementation would probably be to separate the "dirtying of change flags" from "updates" and then after we call `updateState` we clear all the flags. This would for instance allow us to stop calling `updateState` when setting a layer to `visible: false` while still being sure we only update what's needed when `visible` become `true` again.